lib/grid.rb

```diff
@@ -1,24 +1,38 @@
 class Grid
-  def template
-      <<~GRID
-          1  2  3
-         __ __ __
-      A |  |  |  |
-        |__|__|__|
-      B |  |  |  |
-        |__|__|__|
-      C |  |  |  |
-        |__|__|__|
-      GRID
-    end
+  attr_accessor :structure
 
-  def coordinate_valid?(coordinate)
-    coordinates.include?(coordinate)
+  def initialize
+    @structure = grid_structure
+  end
+
+  def mark(coordinate)
+    structure[coordinate] = 'X'
+  end
+  
+  def coordinate_invalid?(coordinate)
+    !structure.key?(coordinate)
+  end
+
+  def get_value(coordinate)
+    structure[coordinate]
+  end
+
+  def current_grid
```

Comment:
I don't think this is the right name for the method, and I think the naming miss embodies the confusion about where to put this and how to take it apart.

This isn't the current grid. It is a template for the grid, with the logic needed to inject the right values built into the template. That is a ton of logic. You aren't allowed to use external libraries for this solution, but check out the library mustache http://mustache.github.io/mustache.5.html for ideas about logicless templates.

Right now this is a super logic-full template, and you need to convert it to logicless in order to separate the grid structure from the template. That is going to mean that you need to work harder to put the values into the template.

Couple of hints to get you on your way. I would expect another printer type object to ask the grid for its value. It could figure out the coordinates from the template itself. For example, if you had a string template `| {A1} |` you could search this string template for anything matching curly braces with a letter. Pulling that coordinate out of the template itself you could then ask the grid for the value and then substitute it.

I would put this logic inside the grid printer.

Reply:
Super helpful, and I figured it out! 🙌